go/coordinator/internal/logservice/grpc/record_log_service_test.go

```diff
@@ -0,0 +1,107 @@
+	"google.golang.org/protobuf/proto"
+	"gorm.io/gorm"
+	"testing"
+)
+
+func encodeVector(dimension int32, vector []float32, encoding coordinatorpb.ScalarEncoding) *coordinatorpb.Vector {
+	buf := new(bytes.Buffer)
+	err := binary.Write(buf, binary.LittleEndian, vector)
+	if err != nil {
+		panic(err)
+	}
+
+	return &coordinatorpb.Vector{
+		Dimension: dimension,
+		Vector:    buf.Bytes(),
+		Encoding:  encoding,
+	}
+}
+
+func GetTestEmbeddingRecords(collectionId string) (recordsToSubmit []*coordinatorpb.SubmitEmbeddingRecord) {
+	testVector1 := []float32{1.0, 2.0, 3.0}
+	testVector2 := []float32{1.2, 2.24, 3.2}
+	testVector3 := []float32{7.0, 8.0, 9.0}
+	recordsToSubmit = make([]*coordinatorpb.SubmitEmbeddingRecord, 0)
+	recordsToSubmit = append(recordsToSubmit, &coordinatorpb.SubmitEmbeddingRecord{
+		Id:           types.NewUniqueID().String(),
+		Vector:       encodeVector(10, testVector1, coordinatorpb.ScalarEncoding_FLOAT32),
+		Operation:    coordinatorpb.Operation_ADD,
+		CollectionId: collectionId,
+	})
+	recordsToSubmit = append(recordsToSubmit, &coordinatorpb.SubmitEmbeddingRecord{
+		Id:           types.NewUniqueID().String(),
+		Vector:       encodeVector(6, testVector2, coordinatorpb.ScalarEncoding_FLOAT32),
+		Operation:    coordinatorpb.Operation_UPDATE,
+		CollectionId: collectionId,
+	})
+	recordsToSubmit = append(recordsToSubmit, &coordinatorpb.SubmitEmbeddingRecord{
+		Id:           types.NewUniqueID().String(),
+		Vector:       encodeVector(10, testVector3, coordinatorpb.ScalarEncoding_FLOAT32),
+		Operation:    coordinatorpb.Operation_ADD,
+		CollectionId: collectionId,
+	})
+	return recordsToSubmit
+}
+
+func resetLogTable(db *gorm.DB) {
+	db.Migrator().DropTable(&dbmodel.RecordLog{})
+	db.Migrator().CreateTable(&dbmodel.RecordLog{})
+}
+
+func TestServer_PushLogs(t *testing.T) {
+	// setup server
+	s, err := New(Config{
+		DBProvider: "postgres",
+		DBConfig:   dbcore.GetDBConfigForTesting(),
+		StartGrpc:  false,
+	})
+	if err != nil {
+		t.Fatalf("error creating server: %v", err)
+	}
+	db := dbcore.GetDB(context.Background())
+	resetLogTable(db)
+
+	// push some records
+	collectionId := types.NewUniqueID()
+	recordsToSubmit := GetTestEmbeddingRecords(collectionId.String())
+	pushRequest := logservicepb.PushLogsRequest{
+		CollectionId: collectionId.String(),
+		Records:      recordsToSubmit,
+	}
+	response, err := s.PushLogs(context.Background(), &pushRequest)
+	assert.Nil(t, err)
+	assert.Equal(t, int32(3), response.RecordCount)
+
+	var recordLogs []*dbmodel.RecordLog
+	db.Where("collection_id = ?", types.FromUniqueID(collectionId)).Find(&recordLogs)
+	assert.Len(t, recordLogs, 3)
+	for index := range recordLogs {
+		assert.Equal(t, int64(index+1), recordLogs[index].ID)
+		assert.Equal(t, collectionId.String(), *recordLogs[index].CollectionID)
+		record := &coordinatorpb.SubmitEmbeddingRecord{}
+		if err := proto.Unmarshal(*recordLogs[index].Record, record); err != nil {
+			panic(err)
```

Comment:
Should not panic here?

Reply:
Can we add a no-panic linter rule? Panic should be rare and we can expect ourselves to call its permissible use out explicitly.

Reply:
even in tests?

Reply:
I just mean orthogonally. Not in tests.

Reply:
Agree, we should almost never panic. I would just assert.NotNil here.